// @metamask/rpc-errors/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.errorCodes = exports.providerErrors = exports.rpcErrors = exports.getMessageFromCode = exports.serializeError = exports.serializeCause = exports.EthereumProviderError = exports.JsonRpcError = void 0;
var classes_1 = require("./classes");
Object.defineProperty(exports, "JsonRpcError", { enumerable: true, get: function () { return classes_1.JsonRpcError; } });
Object.defineProperty(exports, "EthereumProviderError", { enumerable: true, get: function () { return classes_1.EthereumProviderError; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "serializeCause", { enumerable: true, get: function () { return utils_1.serializeCause; } });
Object.defineProperty(exports, "serializeError", { enumerable: true, get: function () { return utils_1.serializeError; } });
Object.defineProperty(exports, "getMessageFromCode", { enumerable: true, get: function () { return utils_1.getMessageFromCode; } });
var errors_1 = require("./errors");
Object.defineProperty(exports, "rpcErrors", { enumerable: true, get: function () { return errors_1.rpcErrors; } });
Object.defineProperty(exports, "providerErrors", { enumerable: true, get: function () { return errors_1.providerErrors; } });
var error_constants_1 = require("./error-constants");
Object.defineProperty(exports, "errorCodes", { enumerable: true, get: function () { return error_constants_1.errorCodes; } });
